File: abbreviations.rb
# encoding: UTF-8
require 'set'
require 'algorithms'
# mod for doing stuff w/ abbrevs
module Abbrev
  # a list of abbreviations culled, at least in part, from the USPS
  # official list https://www.usps.com/send/official-abbreviations.htm#2
  # this might be too country-specific, but it's a start.
  classes = [
    ["alley", "aly"],
    ["and", "&"],
    ["annex", "anx"],
    ["arcade", "arc"],
    ["avenue", "ave"],
    ["beach", "bch"],
    ["burg", "bg"],
    ["bluff", "blf"],
    ["boulevard", "blvd"],
    ["bend", "bnd"],
    ["branch", "br"],
    ["bridge", "brg"],
    ["brook", "brk"],
    ["bottom", "btm"],
    ["bayoo", "byu"],
    ["circle", "cir"],
    ["club", "clb"],
    ["cliff", "clf"],
    ["common", "cmn"],
    ["corner", "cor"],
    ["camp", "cp"],
    ["cape", "cpe"],
    ["crescent", "cr","cres"],
    ["creek", "crk"],
    ["course", "crse"],
    ["crest", "crst"],
    ["causeway", "cswy"],
    ["court", "ct"],
    ["center", "ctr"],
    ["curve", "curv"],
    ["cove", "cv"],
    ["canyon", "cyn"],
    ["dale", "dl"],
    ["dam", "dm"],
    ["drive", "dr"],
    ["doctor", "dr"],
    ["divide", "dv"],
    ["east", "e"],
    ["estate", "est"],
    ["expressway", "expy"],
    ["extension", "ext"],
    ["field", "fld"],
    ["flat", "flt"],
    ["ford", "frd"],
    ["forge", "frg"],
    ["fork", "frk"],
    ["forest", "frst"],
    ["ferry", "fry"],
    ["fort", "ft"],
    ["freeway", "fwy"],
    ["garden", "gdn"],
    ["glen", "gln"],
    ["green", "grn"],
    ["grove", "grv"],
    ["gateway", "gtwy"],
    ["harbor", "hbr"],
    ["hill", "hl"],
    ["hollow", "holw"],
    ["haven", "hvn"],
    ["highway", "hwy"],
    ["inlet", "inlt"],
    ["island", "is"],
    ["junction", "jct"],
    ["knoll", "knl"],
    ["key", "ky"],
    ["lock", "lck"],
    ["lodge", "ldg"],
    ["loaf", "lf"],
    ["light", "lgt"],
    ["lake", "lk"],
    ["lane", "ln"],
    ["landing", "lndg"],
    ["meadow", "mdw"],
    ["mill", "ml"],
    ["manor", "mnr"],
    ["mission", "msn"],
    ["mount", "mt"],
    ["mountain", "mtn"],
    ["motorway", "mtwy"],
    ["neck", "nck"],
    ["north", "n"],
    ["orchard", "orch"],
    ["parkway", "pkwy"],
    ["place", "pl"],
    ["plain", "pln"],
    ["plaza", "plz"],
    ["pine", "pne"],
    ["prairie", "pr"],
    ["port", "prt"],
    ["passage", "psge"],
    ["point", "pt"],
    ["radial", "radl"],
    ["road", "rd"],
    ["ridge", "rdg"],
    ["river", "riv"],
    ["ranch", "rnch"],
    ["row", "row"],
    ["rapid", "rpd"],
    ["rest", "rst"],
    ["route", "rte"],
    ["shoal", "shl"],
    ["shore", "shr"],
    ["skyway", "skwy"],
    ["south", "s"],
    ["summit", "smt"],
    ["spring", "spg"],
    ["square", "sq"],
    ["street", "st"],
    ["station", "sta"],
    ["stravenue", "stra"],
    ["stream", "strm"],
    ["terrace", "ter"],
    ["turnpike", "tpke"],
    ["track", "trak"],
    ["trace", "trce"],
    ["trafficway", "trfy"],
    ["trail", "trl"],
    ["throughway", "trwy"],
    ["tunnel", "tunl"],
    ["union", "un"],
    ["viaduct", "via"],
    ["vista", "vis"],
    ["ville", "vl"],
    ["village", "vlg"],
    ["valley", "vly"],
    ["view", "vw"],
    ["way", "wy"],
    ["well", "wl"],
    ["west", "w"],
    ["crossing", "xing"],
    ["crossroad", "xrd"],

    # Russian abbreviations
    # Copyright (C) 2011-2012 Dmitry Marakasov
    # from https://github.com/AMDmi3/streetmangler/blob/master/lib/locales/ru.cc#L27
    ["улица", "ул"],
    ["площадь", "пл"],
    ["переулок", "пер", "пер-к"],
    ["проезд", "пр-д"],
    ["шоссе", "ш"],
    ["бульвар", "бул", "б-р"],
    ["тупик", "туп"],
    ["набережная", "наб"],
    ["проспект", "просп", "пр-кт", "пр-т"],
    ["тракт", "тр-т", "тр"],
    ["эстакада", "эст"],
    ["район", "р-н"],
    ["микрорайон", "мкр-н", "мк-н", "мкр", "мкрн"],
    ["посёлок", "поселок", "пос"],
    ["деревня", "дер", "д"],
    ["квартал", "кв-л", "кв"],

    # German abbreviations
    ["anschlussstelle", "as"],
    ["an", "a"],
    ["bahnhof", "bf"],
    ["bei", "b"],
    ["bürgermeister", "bgm"],
    ["der", "d"],
    ["den", "d"],
    ["dem", "d"],
    ["evangelische", "ev", "evang"],
    ["evangelischer", "ev", "evang"],
    ["evangelisches", "ev", "evang"],
    ["evangelisch", "ev", "evang"],
    ["fachhochschule", "fh"],
    ["gasse", "g"],
    ["gemeinschaft", "gem"],
    ["gemeinschafts", "gem"],
    ["georg", "gg"],
    ["groß", "gr"],
    ["große", "gr"],
    ["großer", "gr"],
    ["großes", "gr"],
    ["grundschule", "gs"],
    ["gymnasium", "gym", "gymn"],
    ["hauptbahnhof", "hbf"],
    ["hauptschule", "hs"],
    ["hochschule", "hs"],
    ["in", "i"],
    ["johann", "joh"],
    ["johannes", "joh"],
    ["katholische", "kath"],
    ["katholischer", "kath"],
    ["katholisches", "kath"],
    ["katholisch", "kath"],
    ["kindergarten", "kiga"],
    ["kindertagesstätte", "kita"],
    ["klein", "kl"],
    ["kleine", "kl"],
    ["kleiner", "kl"],
    ["kleines", "kl"],
    ["krankenhaus", "kh", "krkh", "krh", "krhs"],
    ["obere", "ob"],
    ["oberer", "ob"],
    ["oberes", "ob"],
    ["platz", "pl"],
    ["realschule", "rs"],
    ["römisch", "röm"],
    ["samtgemeinde", "sg"],
    ["sankt", "st"],
    ["sebastian", "seb"],
    ["straße", "str"],
    ["und", "u","&"],
    ["universität", "uni"],
    ["unterer", "unt","u"],
    ["unteres", "unt","u"],
    ["untere", "unt","u"],
    ["unter", "u"],
    ["vom", "v"],
    ["von", "v"],
    ["weg", "wg"],
    ["zur", "z"],
    ["zum", "z"],
    ["zu", "z"],
    # Swiss German
    ["strasse", "str"],
    
    #compass
     # of course, this is horribly english-specific...
     # but how would one expand this in a sensible fashion to
     # cover other languages?
    ["n", "north"],
    ["e", "east"],
    ["s", "south"],
    ["w", "west"],
    # german
    ["n", "nord"],
    ["o", "ost"],
    ["s", "süd"],
  ]

#build substitution rules out of classes (plz do this only once per redactionbot-start)
@@rules = Hash.new(Set.new)
for clazz in classes
  for elem in clazz
    @@rules[elem] = @@rules[elem] | ((Set.new clazz) - (Set.new [elem]))
  end
end
#special rules like kill spaces, dashes and dots
@@rules[' '] = @@rules[elem] | Set.new(['', '-', '.', '. '])
@@rules['-'] = @@rules[elem] | Set.new([' ', ''])
@@rules['.'] = @@rules[elem] | Set.new([' ', ''])


  # function for expanding a string into a list of strings
    def self.manglenext(heap, manglerules, target)
      if !heap.empty?()
        #remove the best unvisited word from queue and mangle it
        wordstart, wordend = heap.next!()
        print "\n------\n"
        print [wordstart]
        print " - "
        print [wordend]
        #call every rule
        for rule in manglerules.keys()
          #and try to use it
	  if rule == ' '
	    print "\nSPACE\n"
	  end
          #execute rule (just split once!!)
          newsplit = wordend.split(rule,2)
          # if rule doesn't apply len != 2
          if newsplit.size() == 2
            for substitute in manglerules[rule]
              newwordstart = wordstart + newsplit[0] + substitute
              newwordend = newsplit[1]
              #everything in wordstart have to match targets first characters
              if target.start_with?(newwordstart)
                #if we found our string we're happy
		print "\n>>"
		print [target]
		print [newwordstart + newwordend]
                if target == newwordstart + newwordend
                  puts "Found"
                  return true
                end
                heap.push([newwordstart,newwordend],-newwordend.size())
                #to avoid loops with insert space (and insert special rule ' ')
                if rule != ' '
		  newwordspaceend = ' ' + newwordend
                  #if we found our string we're happy
                  if target == newwordstart + newwordspaceend
                    puts "Found"
                    return true
                  end
                  heap.push([newwordstart,newwordspaceend],-newwordend.size()) # insert space
                end
              end
            end
          end
        end
      end
      return false
    end

  # TODO: may need some work for internationalisation
  def self.equal_expansions(a, b)
    input1 = a.downcase() + ' '
    input2 = b.downcase() + ' '
    # TODO: insert abbrev-python-v2 algo here
    if input1 == input2
      #shortcut if string a matches string b
      return true
    end
    # filter rules? maybe if words are long enough / dont remove special rules
    
    #init toextend (priorityqueue)
    extendforwpq = Containers::PriorityQueue.new
    extendforwpq.push(['',input1],0)
    extendbackwpq = Containers::PriorityQueue.new
    extendbackwpq.push(['',input2],0)
    
        
    until extendforwpq.empty?() and extendbackwpq.empty?()
      print "\nForward"
      if manglenext(extendforwpq, @@rules, input2)
        return true
      end
      print "\nBackward"
      if manglenext(extendbackwpq, @@rules, input1)
        return true
      end
    end
    return false
  end

end

File: abbreviations.py
#!/usr/bin/python
# coding: utf-8
#License: Beerware
#Author: MonkZ
import sys
from heapq import heappush, heappop

if not(len(sys.argv) == 3 or (len(sys.argv) == 4 and sys.argv[3] == "-v")):
  print "Please pass two strings to this program"
  exit(1)
target1 = sys.argv[1].decode("utf-8").lower()
target2 = sys.argv[2].decode("utf-8").lower()
verbose = len(sys.argv) == 4

if verbose:
  print "TEST: %s " % target1
  print "TEST: %s " % target2


#classes of strings/abbrvs/synonyms
#we can improve this by check our tag location
#if it is in UK we should use only UK classes and general rules
classes = [
    [u'alley', u'aly'],
    [u'and', u'&'],
    [u'annex', u'anx'],
    [u'arcade', u'arc'],
    [u'avenue', u'ave'],
    [u'beach', u'bch'],
    [u'burg', u'bg'],
    [u'bluff', u'blf'],
    [u'boulevard', u'blvd'],
    [u'bend', u'bnd'],
    [u'branch', u'br'],
    [u'bridge', u'brg'],
    [u'brook', u'brk'],
    [u'bottom', u'btm'],
    [u'bayoo', u'byu'],
    [u'circle', u'cir'],
    [u'club', u'clb'],
    [u'cliff', u'clf'],
    [u'common', u'cmn'],
    [u'corner', u'cor'],
    [u'camp', u'cp'],
    [u'cape', u'cpe'],
    [u'crescent', u'cr', u'cres'],
    [u'creek', u'crk'],
    [u'course', u'crse'],
    [u'crest', u'crst'],
    [u'causeway', u'cswy'],
    [u'court', u'ct'],
    [u'center', u'ctr'],
    [u'curve', u'curv'],
    [u'cove', u'cv'],
    [u'canyon', u'cyn'],
    [u'dale', u'dl'],
    [u'dam', u'dm'],
    [u'drive', u'dr'],
    [u'doctor', u'dr'],
    [u'divide', u'dv'],
    [u'east', u'e'],
    [u'estate', u'est'],
    [u'expressway', u'expy'],
    [u'extension', u'ext'],
    [u'field', u'fld'],
    [u'flat', u'flt'],
    [u'ford', u'frd'],
    [u'forge', u'frg'],
    [u'fork', u'frk'],
    [u'forest', u'frst'],
    [u'ferry', u'fry'],
    [u'fort', u'ft'],
    [u'freeway', u'fwy'],
    [u'garden', u'gdn'],
    [u'glen', u'gln'],
    [u'green', u'grn'],
    [u'grove', u'grv'],
    [u'gateway', u'gtwy'],
    [u'harbor', u'hbr'],
    [u'hill', u'hl'],
    [u'hollow', u'holw'],
    [u'haven', u'hvn'],
    [u'highway', u'hwy'],
    [u'inlet', u'inlt'],
    [u'island', u'is'],
    [u'junction', u'jct'],
    [u'knoll', u'knl'],
    [u'key', u'ky'],
    [u'lock', u'lck'],
    [u'lodge', u'ldg'],
    [u'loaf', u'lf'],
    [u'light', u'lgt'],
    [u'lake', u'lk'],
    [u'lane', u'ln'],
    [u'landing', u'lndg'],
    [u'meadow', u'mdw'],
    [u'mill', u'ml'],
    [u'manor', u'mnr'],
    [u'mission', u'msn'],
    [u'mount', u'mt'],
    [u'mountain', u'mtn'],
    [u'motorway', u'mtwy'],
    [u'neck', u'nck'],
    [u'north', u'n'],
    [u'orchard', u'orch'],
    [u'parkway', u'pkwy'],
    [u'place', u'pl'],
    [u'plain', u'pln'],
    [u'plaza', u'plz'],
    [u'pine', u'pne'],
    [u'prairie', u'pr'],
    [u'port', u'prt'],
    [u'passage', u'psge'],
    [u'point', u'pt'],
    [u'radial', u'radl'],
    [u'road', u'rd'],
    [u'ridge', u'rdg'],
    [u'river', u'riv'],
    [u'ranch', u'rnch'],
    [u'row', u'row'],
    [u'rapid', u'rpd'],
    [u'rest', u'rst'],
    [u'route', u'rte'],
    [u'shoal', u'shl'],
    [u'shore', u'shr'],
    [u'skyway', u'skwy'],
    [u'south', u's'],
    [u'summit', u'smt'],
    [u'spring', u'spg'],
    [u'square', u'sq'],
    [u'street', u'st'],
    [u'station', u'sta'],
    [u'stravenue', u'stra'],
    [u'stream', u'strm'],
    [u'terrace', u'ter'],
    [u'turnpike', u'tpke'],
    [u'track', u'trak'],
    [u'trace', u'trce'],
    [u'trafficway', u'trfy'],
    [u'trail', u'trl'],
    [u'throughway', u'trwy'],
    [u'tunnel', u'tunl'],
    [u'union', u'un'],
    [u'viaduct', u'via'],
    [u'vista', u'vis'],
    [u'ville', u'vl'],
    [u'village', u'vlg'],
    [u'valley', u'vly'],
    [u'view', u'vw'],
    [u'way', u'wy'],
    [u'well', u'wl'],
    [u'west', u'w'],
    [u'crossing', u'xing'],
    [u'crossroad', u'xrd'],
    
    # Russian abbreviations
    # Copyright (C) 2011-2012 Dmitry Marakasov
    # from https://github.com/AMDmi3/streetmangler/blob/master/lib/locales/ru.cc#L27
    [u'улица', u'ул'],
    [u'площадь', u'пл'],
    [u'переулок', u'пер', u'пер-к'],
    [u'проезд', u'пр-д'],
    [u'шоссе', u'ш'],
    [u'бульвар', u'бул', u'б-р'],
    [u'тупик', u'туп'],
    [u'набережная', u'наб'],
    [u'проспект', u'просп', u'пр-кт', u'пр-т'],
    [u'тракт', u'тр-т', u'тр'],
    [u'эстакада', u'эст'],
    [u'район', u'р-н'],
    [u'микрорайон', u'мкр-н', u'мк-н', u'мкр', u'мкрн'],
    [u'посёлок', u'поселок', u'пос'],
    [u'деревня', u'дер', u'д'],
    [u'квартал', u'кв-л', u'кв'],

    # German abbreviations
    [u'anschlussstelle', u'as'],
    [u'an', u'a'],
    [u'bahnhof', u'bf'],
    [u'bei', u'b'],
    [u'bürgermeister', u'bgm'],
    [u'der', u'd'],
    [u'den', u'd'],
    [u'dem', u'd'],
    [u'evangelische', u'ev', u'evang'],
    [u'evangelischer', u'ev', u'evang'],
    [u'evangelisches', u'ev', u'evang'],
    [u'evangelisch', u'ev', u'evang'],
    [u'fachhochschule', u'fh'],
    [u'gasse', u'g'],
    [u'gemeinschaft', u'gem'],
    [u'gemeinschafts', u'gem'],
    [u'groß', u'gr'],
    [u'große', u'gr'],
    [u'großer', u'gr'],
    [u'großes', u'gr'],
    [u'grundschule', u'gs'],
    [u'gymnasium', u'gym', u'gymn'],
    [u'hauptbahnhof', u'hbf'],
    [u'hauptschule', u'hs'],
    [u'hochschule', u'hs'],
    [u'johann', u'joh'],
    [u'johannes', u'joh'],
    [u'katholische', u'kath'],
    [u'katholischer', u'kath'],
    [u'katholisches', u'kath'],
    [u'katholisch', u'kath'],
    [u'kindergarten', u'kiga'],
    [u'kindertagesstätte', u'kita'],
    [u'klein', u'kl'],
    [u'kleine', u'kl'],
    [u'kleiner', u'kl'],
    [u'kleines', u'kl'],
    [u'krankenhaus', u'kh', u'krkh', u'krh', u'krhs'],
    [u'obere', u'ob'],
    [u'oberer', u'ob'],
    [u'oberes', u'ob'],
    [u'platz', u'pl'],
    [u'realschule', u'rs'],
    [u'römisch', u'röm'],
    [u'samtgemeinde', u'sg'],
    [u'sankt', u'st'],
    [u'sebastian', u'seb'],
    [u'straße', u'str'],
    [u'und', u'u', u'&'],
    [u'universität', u'uni'],
    [u'unterer', u'unt'],
    [u'unteres', u'unt'],
    [u'untere', u'unt'],
    [u'unter', u'u'],
    [u'vom', u'v'],
    [u'von dem', u'vd'],
    [u'von der', u'vd'],
    [u'von', u'v'],
    [u'weg', u'wg'],
    [u'zu', u'z'],
    [u'zum', u'z'],
    [u'zur', u'z'],
    # Swiss German
    [u'strasse', u'str'],
  
  # In languages like German that use compound words with no spaces, we will find
  # words that end with an abbreviated suffix. For now let's treat these as a special
  # case to avoid too much extra load.
    [u'weg', u'wg'],
    [u'strasse', u'str'],
    [u'straße', u'str'],
    [u'gasse', u'g'],
    [u'platz', u'pl'],

  # of course, this is horribly english-specific...
  # but how would one expand this in a sensible fashion to
  # cover other languages?
    [u'n'  , u'north'],
    [u'e'  , u'east'],
    [u's'  , u'south'],
    [u'w'  , u'west'],
  ]

#build substitution rules out of classes (plz do this only once per redactionbot-start)
rules = {}
for clazz in classes:
  for elem in clazz:
    if not elem in rules.keys():
      rules[elem] = set([])
    rules[elem] = rules[elem] | (set(clazz) - set([elem]))
    
#filter rules against targets
filteredforwardrules = {}
for rule in rules.keys():
  if target1.find(rule) != -1:
    filteredforwardrules[rule] = rules[rule]

filteredbackwardrules = {}
for rule in rules.keys():
  if target2.find(rule) != -1:
    filteredbackwardrules[rule] = rules[rule]

#add special rules like "kill spaces"
filteredforwardrules[u' '] = set([u' ', u'', u'-', u'.', u'. '])
filteredforwardrules[u'-'] = set([u'-', u' '])
filteredforwardrules[u'.'] = set([u'.', u' ', u''])
filteredbackwardrules[u' '] = set([u' ', u'', u'-', u'.', u'. '])
filteredbackwardrules[u'-'] = set([u'-', u' '])
filteredbackwardrules[u'.'] = set([u'.', u' ', u''])


if verbose:
  print filteredforwardrules
  print filteredbackwardrules

#rulemangling
toextendforw = [(0,('',target1))]
toextendbackw = [(0,('',target2))]

def manglenext(heap, manglerules, target):
  if heap != []:
    #remove the best unvisited word from queue and mangle it
    wdist, (wordstart, wordend) = heappop(heap)
    #if we found our string we're happy
    if wordstart + wordend == target:
      print "Found"
      exit(0)

    if verbose:
      if target == target1:
        print "Forw:"
      else:
        print "Backw:"
      print "pop %s | %s - dist: %i" % (wordstart, wordend, wdist)

    #call every rule
    for rule in manglerules.keys():
      #and try to use it (maybe this could be improved by find our ruletrigger in first place)
      for substitute in manglerules[rule]:
        #execute rule (just split once!!)
        newsplit = wordend.split(rule,1)
        # if rule doesn't apply len != 2
        if len(newsplit) == 2:
          newwordstart = wordstart + newsplit[0] + substitute
          newwordend = newsplit[1]
          #everything in wordstart have to match targets first characters
          if target.startswith(newwordstart):
            heappush(heap, (len(newwordend),(newwordstart,newwordend)))
            #to avoid loops with insert space (and insert special rule ' ')
            if rule != ' ':
              heappush(heap, (len(newwordend),(newwordstart,' '+newwordend))) # insert space

#try rules on every string until we've no more strings
while(toextendforw != [] or toextendbackw != []):
  manglenext(toextendforw, filteredforwardrules, target2)
  manglenext(toextendbackw, filteredbackwardrules, target1)

# :(
print "NOT Found"
exit(2)
